### PR TITLE
Reshape UMD to respect module/export first.

### PR DESCRIPTION
**Pull request by leebyron**
Ran into a problem with the UMD wrapper. Here's the scenario:

We wrap libraries in UMD to make them easier to deploy in any environment, however one common environment is CJS to AMD-compiler.

If a library is exported as UMD, but treated as CJS in a local environment, and then compiled to AMD (via a wrapper), then our UMD wrapper misidentifies the environment as AMD, when we want it to behave as CJS. define() is called from within define(), and the outer-most define() ends up thinking its module exports nothing, causing sad trombone music to play out of it's tiny 8-bit speakers.

In other words: AMD defines both "module"/"exports" and "define" where CJS defines only "module"/"exports". However, AMD environments do not globally define "module"/"exports" - only locally to a module definition, in which case they should be treated as CJS treats them.

What this means is that we actually want to detect "module"/"exports" first, falling back to detecting "define" (assuming a global level).

### CODE REVIEW

**Comment by leebyron** [comment]
Here's an example of what a common cjs->amd looks like:

```js
// umd_module.js

(function (global, factory) {
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  typeof exports === 'object' ? factory(exports) :
  (global.myModule = {}, factory(global.myModule))
}(this, function (exports) { 'use strict';

   exports.foo = 'bar';

}));


// umd_module_amd_wrapped.js

define('myModule', ['exports'], function (exports) {

	(function (global, factory) {
    typeof define === 'function' && define.amd ? define(['exports'], factory) :
    typeof exports === 'object' ? factory(exports) :
    (global.myModule = {}, factory(global.myModule))
  }(this, function (exports) { 'use strict';

     exports.foo = 'bar';

  }));

});
```

Note how in the amd transformed source, `define` would be detected when the `exports` behavior is desired.

**Comment by Rich-Harris** [comment]
Oh no, we're in [Trek Glowacki's emo hellscape](https://medium.com/@trek/last-week-i-had-a-small-meltdown-on-twitter-about-npms-future-plans-around-front-end-packaging-b424dd8d367a)! The 'official' UMD pattern puts the AMD check first, because of [reasons](https://github.com/umdjs/umd/pull/22) (I've experienced this myself, though I don't remember the specifics).

It seems like the problem here is the part that's mistakenly wrapping the module as AMD, ignoring the fact that it's already a valid AMD module. Can that be fixed? Or is there a more aggressive check we could do (e.g. testing for `global.exports` instead of `exports`)?

**Comment by leebyron** [comment]
Ugghhh.

I'm not sure `global.exports` circumvents the tooling issue since it might not be global.

I wonder if a check for both exports and module would suffice?

`typeof exports === 'object' && typeof module === 'object'`

**Comment by leebyron** [comment]
Reading more about this is causing me to consider abandoning UMD in favor of two separate files - one expressly designed to be consumed by CJS (in which the AMD wrapper would not break anything) and one to be loaded in directly as a file in a <script> tag.

**Comment by Rich-Harris** [comment]
I've had to resort to similar tactics in the past. It's a real PITA.

**Comment by leebyron** [comment]
FWIW, React also distributes in UMD, but checks for CJS first. I followed the chain and this is browserify "standalone" behavior which is provided by https://github.com/ForbesLindesay/umd/blob/master/template.js

**Comment by eventualbuddha** [comment]
Not to simply appeal to an authority, but I believe that browserify probably has a fair amount of road paved in the right direction and that emulating it for the umd wrapper is probably a good idea. I'm in favor of the switch.

**Comment by leebyron** [comment]
Yeah... I think UMD is ultimately flawed no matter how you go about it, but I can say two things:

1) In the time I've worked on React, no one has complained about the UMD wrapper doing the wrong thing. The last few months Immutable has been in production, it's CJS-first UMD wrapper solicited no complaints.

2) When I swapped to the AMD-first UMD wrapper in esperanto - I ran into this issue pretty quickly :)

**Comment by Rich-Harris** [comment]
> Not to simply appeal to an authority

In a situation like this, where there's no right answer, I guess authority (and recent experience) is all we *can* appeal to. So let's try CJS-first. @leebyron you predicted correctly - this won't auto-merge, can you update this branch? Thanks

**Comment by leebyron** [comment]
Done!

**Comment by Rich-Harris** [comment]
Nice! Will cut a new release later today

**Comment by leebyron** [comment]
Thanks, let me know when you do! I'll repackage and release Immutable :)

**Comment by Rich-Harris** [comment]
@leebyron 0.6.0 is ready to go
